### DISCUSSION TITLE
CYD as Serial Monitor

### DISCUSSION THREAD

**Discussion by tomvanderputte** [witnessmenow/ESP32-Cheap-Yellow-Display#113] [Q&A] [answered]
My goal is to use the CYD (powered by a powerbank) as a serial monitor for arduino pro mini (3.3V) sensors, so I can debug them without having to attach them to my PC.

 For testing:  
 I have a simple script on both devices that work well independently when connected to my PC.
Connected the pro mini via FTDI->USB: it sends out the P character (visible on the serial monitor) 
Connected the CYD via USB: when I type the letter P, it flashes the led and displays it on the screen
So the separate parts work.

In practice:
I've connected the CYD P1 port RX (IO3) to the arduinos TX, and connected the P1 GND to Arduino GND.
But nothing happens, at all. I thought it maybe had to do with the display stuff, but it looks as if the RX on the CYD is simply not oing anything.

Then I saw this:
https://github.com/witnessmenow/ESP32-Cheap-Yellow-Display/discussions/75
Tried that with serial2, but no luck either.

Then I tried it the other way around: 
Send something from CYD to Arduino, switched CYD TX to Arduino RX and that works.

Questions:

1. Would the P1 port still work if connected through USB at the same time to the PC (so that would make 2 serial connections on the same pins)?
2. Is the logic level on the P1 port 3V or 5V (3V on the RX should work, even with a 5V logic level I think, since these are normally quite tolerant for lower voltages)?
3. Does anybody have any clues on what's going on here? 
4. Could the RX pin be fried , and how could I test this?

**Comment by tomvanderputte** [comment, marked as answer]
Update:
I tried using the FTDI Adapter to see if I could connect the PC serial monitor to the P1 port. It doesn't work on 3V (Vin needs 5V). But I was not sure if 5V would fry the RX pin. Tried it anyway, and typing in any character on the serial monitor immediately showed up on the display.

So is it safe to assume the RX needs 5V?

**Comment by chetkelly** [comment]
If you have the USB connected to the CYD it will hold the RX line high and you will not receive anything on that connector, might that explain what you were seeing?

**Comment by tomvanderputte** [comment]
Well not sure, would that also be the case if usb is connected to a powerbank?

The thing that worked, described (not so clearly) above was:

Powerbank via usb to cyd
Pc -> usb -> FTDI(5V)-RX -> TX-CYD
Pc -> usb -> FTDI(5V)-TX -> RX-CYD

This allowed me to receive stuff from the serial monitor.
I also tried it with the FTDI selector on 3V , and that didn't work.

So all the above led me to believe I need 5v on the cyd uart, and as such i would need a logic lvl converter ?

**Reply by chetkelly** [reply]
Once the USB has power the usb<->uart IC will be powered up so it should still apply with the powerbank.

The rx/tx pins are (iirc) connected directly to the ESP32 so should be 3.3v only.

When connected to the FTDI how was the CYD powered?

For your use case you'll want the powerbank supplying 5v directly to P1

**Reply by tomvanderputte** [reply]
On your question on the power, I might have powered it on the p1 port and not by powerbank->usb. Uhmm. Will follow this up with some test scenarios and report back.

**Comment by tomvanderputte** [comment]
And also tested it with an Arduino one at 5V (forgot all about that haha)
![PXL_20240113_204235516](https://github.com/witnessmenow/ESP32-Cheap-Yellow-Display/assets/65624141/e0598759-1465-46dc-a559-64bee08b6e8f)

**Reply by chetkelly** [reply]
If you attach the micro USB to the CYD there it'll stop receiving.

**Reply by Davidowski4** [reply]
Can you share the code please? it would be very useful to me.

**Reply by paelzer** [reply]
@tomvanderputte

Hello

I'm trying to also get this up and running, so far without success. Could you share your code?

BR
Boris

**Comment by simo5x5** [comment]
Hello everyone, 
im using the esp32 CYD (2 USB version, 2.8 display) for a radar project. The P5 connector with Vin, TX,RX and ground doesnt works. I was able to use the CN1 connector (next to the sd card reader) as serial port. This connector has 3.3v,IO27,IO22 and ground. 
Note :  IO22 is shared with the P3 connector.  

Sample code : 
HardwareSerial RadarSerial(2);
RadarSerial.begin(9600, SERIAL_8N1, 27, 22);

After you will be able to use all the standard commands of Serial. Hope this will be helpful, 
Best Regards 
Simo.

**Reply by jonsummers** [reply]
**THIS!  SIMO SOLVED IT!**  

Finally a solution to make the CYD an all-in-one SerialSniffer: 
power up with powerbank and plug into Serial TX of other ESP32.

A humble suggestion:
swap the RX and TX pin defines to match the connector cabling match P5:
reading left to right across connectors: red: pwr, yellow tx, blue rx, black gnd
P5:   "VIN, TX, RX, GND"   is 5V, GPIO1, GPIO3, GND
CN1: "3v3, IO27, IO22, GND" becomes
CN1: (3v3), TX2, RX2, GND 
so TX2 is pin 27, RX2 is pin 22:

_RadarSerial.begin(9600, SERIAL_8N1, 27, 22);_
becomes
_RadarSerial.begin(9600, SERIAL_8N1, 22, 27);_

How to use Serial2 (CN1) as RX and still use USB for Serial.print() debug...

 `const byte TX2 = 27; const byte RX2 = 22; // Note swapped pin assignments`
`HardwareSerial  SerialCN1(2);`
`#define SerialUSB Serial`
`void setup() {`
`  SerialUSB(baudUSB);   // Serial.println() --> USB  equiv Serial(baudUSB, SERIAL_8N1, 3, 1);`
`  SerialCN1(baudCN1, SERIAL_8N1, RX2, TX2);     // Serial2.read()  from CN1`
`}`
`void loop() {  while (SerialCN1.available()) { byte data = SerialCN1.read(); if (data > 31)  SerialUSB.print((char)data); } }`

**Reply by jonsummers** [reply]
Also IO27 is pulled up to 3v3 via 10KΩ. 
For RX sniffing only, probably wiser to use floating IO22 as RX:

![CN1asSerial2git](https://github.com/user-attachments/assets/2faf7d5a-638c-4aad-8d93-5771df323205)

**Reply by simo5x5** [reply]
This was an old project but you are right, better to switch pins. Well done!
